src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -479,6 +496,41 @@ void ReplicatedWriteLog<I>::aio_write(Extents &&image_extents,
                                       bufferlist&& bl,
                                       int fadvise_flags,
                                       Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  if (RWL_VERBOSE_LOGGING) {
+    ldout(cct, 20) << "aio_write" << dendl;
+  }
+  utime_t now = ceph_clock_now();
+  m_perfcounter->inc(l_librbd_rwl_wr_req, 1);
+
+  ceph_assert(m_initialized);
+  {
+    std::shared_lock image_locker(m_image_ctx.image_lock);
+    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+      on_finish->complete(-EROFS);
+      return;
+    }
+  }
+
+  if (ExtentsSummary<Extents>(image_extents).total_bytes == 0) {
+    on_finish->complete(0);
+    return;
+  }
+
+  auto *write_req =
+    new C_WriteRequestT(*this, now, std::move(image_extents), std::move(bl), fadvise_flags, on_finish);
+  m_perfcounter->inc(l_librbd_rwl_wr_bytes, write_req->image_extents_summary.total_bytes);
+
+  /* The lambda below will be called when the block guard for all
+   * blocks affected by this write is obtained */
+  GuardedRequestFunctionContext *guarded_ctx =
+    new GuardedRequestFunctionContext([this, write_req](GuardedRequestFunctionContext &guard_ctx) {
+      write_req->blockguard_acquired(guard_ctx);
+      alloc_and_dispatch_io_req(write_req);
+    });
+
+  detain_guarded_request(GuardedRequest(write_req->image_extents_summary.block_extent(),
```

Comment:
OK, we will give C_BlockIORequest pointer to `detain_guarded_request()`. This decrease the code redundancy of grabbing.

Reply:
Yes, we can give a `C_BlockIORequest` pointer. @yison in later PR we need to add a parameter called is_barrier for flush request.

Reply:
C_BlockIORequest a template, not a class. If we use  C_BlockIORequest pointer, we need to make GuardedRequest become a template too.  But I don't think it's a good idea to make GuardedRequest  become a template due to C_BlockIORequest pointer.

Reply:
We can give C_BlockIORequest  as the parameter of detain_guarded_request(), and create GuardedRequest inside so that the codes will not be redundant.

Reply:
It would be interesting to know, given my other comments about proper encapsulation, if those really need to be templates. For example, could they just take a pointer to a simple interface for all the hooks they need back into RWL?

Reply:
Sorry @dillaman, I don't quite understand your concern. In fact, I just had a check. If we transfer the pointer of C_BlockIORequest to GuardedRequest's constructor, the constructor needs to be a template function and the GuardedRequest doesn't need to be a template class. 
```cpp
class GuardedRequest {
  ....
  template<typename T> GuardedRequest(C_BlockIORequest<T> *p) {
    ...
  }
}
```